pkg/syncer/resize_reconciler.go

```diff
@@ -109,7 +110,11 @@ func newResizeReconciler(
 		UpdateFunc: rc.updatePVC,
```

Comment:
What happens when the TKG PVC's request and capacity is updated to say 10GB but the SV PVC still has the capacity as 5GB and FileSystemResizePending condition, and metadata syncer in TKG restarts? Should we not be registering for PVC `add` event as well?

Reply:
Checked this with Xing. 
Add event is only for creating a new PVC so it will not help here.  When TKG restarts, the object should be picked up as it is still in the queue and be processed again.

Reply:
I did not follow. The resize reconciler runs in syncer. Lets take a case TKG PVC/PV is 1GB and there is no resize pending condition on TKG PVC and SV PVC. Now say the syncer is down. During this time, the TKG PVC is expanded to 2GB. So, the SV PVC is expanded to 2GB(with resize pending) and TKG PVC has resize pending condition. The TKG PV is 2GB post expand. Now you deploy a pod in TKG cluster, so the resize pending condition is removed from TKG PVC. So, you basically have a TKG PVC with no resize pending, but SV PVC with resize pending condition. How are we removing the resize pending condition on SV cluster when the syncer starts to run now? 

If we listen to TKG PVC add event, we can process it and update such SV PVC(to remove resize pending condition). So, the question I asked was - don't we need to register event handler for TKG PVC `add` events. If no, can you help me understand how the resize pending condition will be removed from the SV PVC?

Reply:
The resize reconciler has a resyncPeriod so lost event should be processed again.
https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/syncer/resize_reconciler.go#L90

Also I don't see how adding the "add event" can solve the problem as you can't expand the PVC before the PVC is created (when add event is received).

Reply:
@SandeepPissay Please see Xing's update.

Reply:
There is an assumption made in the code here that we need to listen to only "updatePVC" event on TKG PVC which is incorrect. We could have TKG PVC's FileSystemResizePending condition removed when syncer container is down. In this specific case, when the syncer comes back up again, it would never know about the TKG PVC's FileSystemResizePending condition removal at all as there are no more update events on that TKG PVC(syncer missed this event). In this case, the SV PVC will continue to have "FileSystemResizePending" since resize reconciler totally missed the event on the TKG PVC as its just listening to update event and not add. This is what I'm highlighting as a bug in the code.

I'll repro this somehow to simulate syncer missing the update event and get back.

Reply:
I could repro the problem. Its hard to repro a case where syncer is down when FileSystemResizePending was removed from a TKG PVC. So, I had to orchestrate various events to simulate this particular bug. Note that this is quite possible in customer env.

Steps:
1. Create TKG PVC with 1 GB. This creates a SV PVC with 1GB.
2. Expand the TKG PVC from 1 GB to 2 GB. Both TKG PVC and SV PVC has FileSystemResizePending condition.
3. Disable TKG Controller Manager(aka GCM) in SV cluster so that we can edit PVCSI deployment in TKG cluster to simlate syncer down scenario.
4. Simulate syncer container down in PVCSI by editing PVCSI deployment with non existing image for syncer.
5. Create a Pod in TKG that uses the TKG PVC. Observe that the TKG PVC has FileSystemResizePending condition removed. This is as expected. Note that the FileSystemResizePending is still present in SV PVC since syncer is not running(we disabled it in step 4).
6. Edit PVCSI deployment and provide the right image for the syncer container. Observe that the syncer pod comes up successfully.
7. FileSystemResizePending on SV PVC is never cleared since syncer missed the TKG PVC update event(FileSystemResizePending remove event).

--------------- TKG PVC after step 2 (TKG PVC has FileSystemResizePending as there are no pods deployed yet) -------  spec:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: 2Gi
    storageClassName: vsan-default-storage-policy
    volumeMode: Filesystem
    volumeName: pvc-bf9fa821-9b80-427c-b97f-e086e3a45348
  status:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 1Gi
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2021-01-11T23:44:15Z"
      message: Waiting for user to (re-)start a pod to finish file system resize of
        volume on node.
      status: "True"
      type: FileSystemResizePending
    phase: Bound
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

--------- SV PVC after step 2 (SV PVC has FileSystemResizePending since the TKG PVC was not used by any pod yet) -----
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi
  storageClassName: vsan-default-storage-policy
  volumeMode: Filesystem
  volumeName: pvc-a7ba3380-c45f-466e-80b0-17d7ee882be2
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-01-11T23:44:15Z"
    message: Waiting for user to (re-)start a pod to finish file system resize of
      volume on node.
    status: "True"
    type: FileSystemResizePending
  phase: Bound

--------- TKG PVC (after step 5) --------------
  spec:
    accessModes:
    - ReadWriteOnce
    resources:
      requests:
        storage: 2Gi
    storageClassName: vsan-default-storage-policy
    volumeMode: Filesystem
    volumeName: pvc-bf9fa821-9b80-427c-b97f-e086e3a45348
  status:
    accessModes:
    - ReadWriteOnce
    capacity:
      storage: 2Gi
    phase: Bound
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""

---------- SV PVC (wait for 15 mins after step 7). The FileSystemResizePending is never cleared! -----------------
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi
  storageClassName: vsan-default-storage-policy
  volumeMode: Filesystem
  volumeName: pvc-a7ba3380-c45f-466e-80b0-17d7ee882be2
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-01-11T23:44:15Z"
    message: Waiting for user to (re-)start a pod to finish file system resize of
      volume on node.
    status: "True"
    type: FileSystemResizePending
  phase: Bound

------------------ Syncer logs --------------------------------------------------
{"level":"info","time":"2021-01-11T23:51:34.916865217Z","caller":"syncer/resize_reconciler.go:152","msg":"Resize reconciler: Start","TraceId":"d361c207-1c1b-41d6-8246-834af1465c58"}
{"level":"info","time":"2021-01-11T23:51:34.934884251Z","caller":"syncer/util.go:133","msg":"getPVCKey: PVC key ns-1/1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bf9fa821-9b80-427c-b97f-e086e3a45348","TraceId":"c66e5840-1f1f-4bec-86a5-e5c53c9379fe"}
{"level":"info","time":"2021-01-11T23:51:34.938014784Z","caller":"syncer/volume_health_reconciler.go:111","msg":"addPVC: add ns-1/1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bf9fa821-9b80-427c-b97f-e086e3a45348 to claim queue","TraceId":"c66e5840-1f1f-4bec-86a5-e5c53c9379fe"}
{"level":"info","time":"2021-01-11T23:51:34.938658876Z","caller":"syncer/util.go:133","msg":"getPVCKey: PVC key ns-1/www-web-0","TraceId":"ad52aece-2b35-400f-9f23-bf067ad523ee"}
{"level":"info","time":"2021-01-11T23:51:34.985070119Z","caller":"syncer/volume_health_reconciler.go:111","msg":"addPVC: add ns-1/www-web-0 to claim queue","TraceId":"ad52aece-2b35-400f-9f23-bf067ad523ee"}
{"level":"info","time":"2021-01-11T23:51:34.985199493Z","caller":"syncer/util.go:133","msg":"getPVCKey: PVC key ns-1/www-web-2","TraceId":"ce7334ab-dae9-4194-bd74-ebc7e2f975d8"}
{"level":"info","time":"2021-01-11T23:51:34.985267379Z","caller":"syncer/volume_health_reconciler.go:111","msg":"addPVC: add ns-1/www-web-2 to claim queue","TraceId":"ce7334ab-dae9-4194-bd74-ebc7e2f975d8"}
{"level":"info","time":"2021-01-11T23:51:35.008535928Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bf9fa821-9b80-427c-b97f-e086e3a45348","TraceId":"328272ff-88b7-4904-ac86-ea6f3954eed7"}
{"level":"info","time":"2021-01-11T23:51:35.008652446Z","caller":"syncer/volume_health_reconciler.go:251","msg":"Found Tanzu Kubernetes Grid PV pvc-bf9fa821-9b80-427c-b97f-e086e3a45348 for Supervisor Cluster PVC ns-1/1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bf9fa821-9b80-427c-b97f-e086e3a45348","TraceId":"328272ff-88b7-4904-ac86-ea6f3954eed7"}
{"level":"info","time":"2021-01-11T23:51:35.009487615Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"baf81e2a-6b04-4ad6-97fe-9115cfd362cf"}
{"level":"info","time":"2021-01-11T23:51:35.010217574Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"36d5d7cb-1f8f-4187-ac3a-d6e12bc44bb8"}
{"level":"info","time":"2021-01-11T23:51:35.045953571Z","caller":"syncer/volume_health_reconciler.go:232","msg":"Updated Tanzu Kubernetes Grid PVC for PV pvc-bf9fa821-9b80-427c-b97f-e086e3a45348","TraceId":"328272ff-88b7-4904-ac86-ea6f3954eed7"}
{"level":"info","time":"2021-01-11T23:51:35.098332565Z","caller":"syncer/pvcsi_fullsync.go:95","msg":"FullSync: Creating CnsVolumeMetadata 1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bbbba592-d828-4eb1-9668-576c342c5922 on the supervisor cluster for entity type \"POD\"","TraceId":"3c1ae03b-16e4-49ea-abf3-859b964f0b55"}
{"level":"info","time":"2021-01-11T23:51:35.290037948Z","caller":"syncer/pvcsi_fullsync.go:124","msg":"FullSync: End","TraceId":"3c1ae03b-16e4-49ea-abf3-859b964f0b55"}
{"level":"info","time":"2021-01-11T23:51:36.010211126Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"a2821428-a469-4a12-8ebb-d1211b3e3867"}
{"level":"info","time":"2021-01-11T23:51:36.010828307Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"18a90363-a200-4250-86e5-340f692e2b7f"}
{"level":"info","time":"2021-01-11T23:51:38.010958768Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"3624bd8d-e53c-471c-9071-13e7b0cf50a9"}
{"level":"info","time":"2021-01-11T23:51:38.011502748Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"91b09c50-5ae6-400e-8a08-463a83c22b90"}
{"level":"info","time":"2021-01-11T23:51:42.011703297Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"de719e01-7953-439d-b28d-ad223047ced5"}
{"level":"info","time":"2021-01-11T23:51:42.012165677Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"5f518d2b-1295-467c-95bd-0b9977612c60"}
{"level":"info","time":"2021-01-11T23:51:50.019343897Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"338a607a-d017-4a2b-94d8-20e8f75116f1"}
{"level":"info","time":"2021-01-11T23:51:50.021783206Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"ecefbec0-41e6-440f-9585-2eaff057530b"}
{"level":"info","time":"2021-01-11T23:52:06.022835133Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"707125eb-1ac2-44cc-a8f0-0f21adb6e88d"}
{"level":"info","time":"2021-01-11T23:52:06.029417719Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"3ae0c3c2-76d0-4924-8c09-3de0407cc7f1"}
{"level":"info","time":"2021-01-11T23:52:38.025608856Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"4a58084c-036d-48ee-8cc8-ff598f57aed3"}
{"level":"info","time":"2021-01-11T23:52:38.0318002Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"3e5335ed-83e3-43ec-b895-63afca630398"}
{"level":"info","time":"2021-01-11T23:53:42.100179464Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"c1e56250-17fd-49ba-a28c-1b70b069e054"}
{"level":"info","time":"2021-01-11T23:53:42.115562898Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"e7966761-2386-4d0b-9dc8-ffcf9c0211a1"}
{"level":"info","time":"2021-01-11T23:55:50.107467747Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"f7c89ec6-a527-4e9a-8a30-3cf1be4ac729"}
{"level":"info","time":"2021-01-11T23:55:50.119961081Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"580bbf52-babc-42b7-9558-6a6269e1620a"}
{"level":"info","time":"2021-01-12T00:00:06.115086801Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"22a0d1af-a4f9-47db-9847-55d711f95577"}
{"level":"info","time":"2021-01-12T00:00:06.125467807Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"8c3f58ad-c363-4e4f-974a-fdbe92653ded"}
{"level":"info","time":"2021-01-12T00:05:06.123062632Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"09330f7e-7c03-421d-a1cd-c97eb65e6ebf"}
{"level":"info","time":"2021-01-12T00:05:06.134907275Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"97a83ba5-eba0-450d-9fa8-dac5ac695287"}
{"level":"info","time":"2021-01-12T00:10:06.138092273Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-0","TraceId":"e544b957-043e-4ded-b59f-15a8fbbc6b97"}
{"level":"info","time":"2021-01-12T00:10:06.165532356Z","caller":"syncer/volume_health_reconciler.go:209","msg":"syncPVC: Found Supervisor Cluster PVC: ns-1/www-web-2","TraceId":"d229f142-5bca-4b4e-8c5f-0035d5036be8"}

Reply:
How long have you waited before verifying FileSystemResizePending on SV PVC is still not cleared?
I still don't see how a "Add event" can fix this problem.

Reply:
Now it has been 147 minutes and the FileSystemResizePending condition on SV PVC is not yet cleared.

```
root@4221f33ed3730a0d3ce39b629f58d36a [ ~ ]# kubectl -n ns-1 get pvc 1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bf9fa821-9b80-427c-b97f-e086e3a45348
NAME                                                                        STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS                  AGE
1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bf9fa821-9b80-427c-b97f-e086e3a45348   Bound    pvc-a7ba3380-c45f-466e-80b0-17d7ee882be2   1Gi        RWO            vsan-default-storage-policy   147m

root@4221f33ed3730a0d3ce39b629f58d36a [ ~ ]# kubectl -n ns-1 get pvc 1cf700d9-63ab-4c75-a8ac-2d86ec692c34-bf9fa821-9b80-427c-b97f-e086e3a45348 -o yaml
........
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 2Gi
  storageClassName: vsan-default-storage-policy
  volumeMode: Filesystem
  volumeName: pvc-a7ba3380-c45f-466e-80b0-17d7ee882be2
status:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 1Gi
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2021-01-11T23:44:15Z"
    message: Waiting for user to (re-)start a pod to finish file system resize of
      volume on node.
    status: "True"
    type: FileSystemResizePending
  phase: Bound

```

> I still don't see how a "Add event" can fix this problem.

Its simple. You need to listen on PVC add event too. When the syncer starts, it gets the add event for the TKG PVC and you can queue processing this TKG PVC in resize reconciler queue. Currently you are not listening on the PVC add event, so you are missing processing this TKG PVC and ending up with the bug I described. Btw, I raised this concern on Nov 17th in this code, I also provided proof of this bug today by reproducing the bug. Not sure what else I need to do :)

Reply:
Liping will try to reproduce this.

Reply:
@SandeepPissay The purpose of this PR is to fix  "initResizeReconciler in metadatasyncer.go does not handle WaitForCacheSync errors." I need to check in  this part since the bug for this PR is a csi-u2-blocker.
The problem you mentioned is out of the scope of this bug and I plan to file a new bug to track the issue you mentioned.

Reply:
Per offline discussion, I have filed a new bug to track the problem that Sandeep pointed out and will be fixed later in a separate PR.

Reply:
Do you have an issue filed in GitHub? If no, can you file one and also add a link here to fix it?